lib/launcher.rb

```diff
@@ -149,34 +173,115 @@ def write_pid
       retry
     end
 
-    def webrick_server(app, addresses, port)
+    def add_puma_server_callback(sd_notify)
+      events = ::Puma::Events.new(::Proxy::LogBuffer::Decorator.instance, ::Proxy::LogBuffer::Decorator.instance)
+      events.register(:state) do |status|
+        if status == :running
+          logger.debug "Finished launching a puma instance, #{sd_notify.pending} instances to go..."
+          sd_notify.ready_all { logger.info("Smart proxy has finished launching #{sd_notify.total} puma instances, ready to serve requests.") }
+        end
+      end
+      events
+    end
+
+    def add_puma_server(app, address, port, conn_type, sd_notify)
+      logger.debug "Launching Puma listener at #{address} port #{port}"
+      if conn_type == :ssl
+        require 'cgi'
+        query_list = app[:SSLArgs].to_a.map do |x|
+          "#{CGI::escape(x[0].to_s)}=#{CGI::escape(x[1])}"
+        end
+        host = "ssl://#{address}/?#{query_list.join('&')}"
```

Comment:
As @iNecas has reported, if `address` contains `:` or `.` characters, this is being misparsed.  Using `ssl://[#{address}]/` is probably the right thing to use.

Reply:
This is a common code for both IPv4 and IPv6, we feed it correctly: `['0.0.0.0', '[::1]']` so I guess this will not fix the issue.